WIP repl: emit uncaughtException

BridgeAR opened the pull request:
This was just a rough idea after looking into #19998.

I have no real opinion if this is a good idea or not and though it might be good if others give their opinion. This way it seems like we would be able to actually report `uncaughtException` if a listener is attached.

If this is indeed something we could do, I'll also add tests.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented on lib/repl.js:
```diff
@@ -416,7 +418,9 @@ function REPLServer(prompt,
-    if (isError && e.stack) {
+    if (process.listenerCount('uncaughtException') !== 0) {
+      process.emit('uncaughtException', e);
+    } else if (isError && e.stack) {
```
I don’t think we want the event being emitted and information being written to stderr to be mutually exclusive? I’d generally expect that the REPL shows global errors that aren’t caught, even if there is an `uncaughtException` listener…

apapirovski commented:
This works in the opened `repl` but one of the main use cases domains serve in `repl` is in its programmatic usage where multiple `repl` instances can be created. This would pretty badly break that for anyone using `uncaughtException`.

BridgeAR commented:
Is it actually possible to detect if the `repl` is opened or not? In that case we could just do it there?

AyushG3112 commented:
The question which comes to my mind is whether we should allow one REPL session a look into `uncaughtException`s of another REPL session. I'd personally be against that as it will lead to unintended situations which would be potentially confusing to users.

BridgeAR commented:
Is there interest in trying to pursue this further? I think it would make sense in case for the opened `repl` in case we can distinguish that from the programmatic usage. I just do not know if we have that information / if it is possible to get that information.

addaleax commented:
@bridgear We do make some customizations for the Node.js main REPL, yes – `lib/internal/repl.js` is the one we use for that, rather than `lib/repl.js`

(Would also be okay with that, yes)

refack commented:
Can one of the admins verify this patch?

BridgeAR commented:
PTAL

I just pushed another commit that limits this functionality strictly to the very first opened repl. In case a second one is opened, it will deactivate the behavior for all.

BridgeAR commented:
@addaleax @apapirovski @AyushG3112 what do you think about the approach?

addaleax commented on lib/repl.js:
```diff
@@ -347,7 +351,8 @@ function REPLServer(prompt,
           process.domain.emit('error', err);
-          process.domain.exit();
+          if (process.domain)
+            process.domain.exit();
```
Can you explain why these changes became necessary? It’s not obvious from looking at the diff…

BridgeAR replied:
When emitting `uncaughtException` while being in an active domain, the domain will be set to null. That results in error here. I worked around it now by just emitting on next tick.

addaleax commented on lib/repl.js:
```diff
@@ -158,6 +160,8 @@ function REPLServer(prompt,
     options = {};
   }
 
+  instances++;
```
I think I’d prefer it if we could do this only for the built-in REPL rather than counting instances?

BridgeAR replied:
Done

jasnell commented:
Ping @BridgeAR ... what's the status on this?

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/19139/

PTAL. I guess this is actually getting somewhere. I am just not to sure how to write a test for this.

antsmartian commented on lib/repl.js:
```diff
@@ -436,19 +437,28 @@ function REPLServer(prompt,
+        process.listenerCount('uncaughtException') !== 0) {
+      process.nextTick(() => {
+        process.emit('uncaughtException', e);
+        top.clearBufferedCommand();
```
nit: These lines : 

```
      top.clearBufferedCommand();
      top.lines.level = [];
      top.displayPrompt();
```

are repeated in both the blocks.

BridgeAR replied:
Well, I could create a function to execute them but it felt like being overkill to be so DRY here (the scope is different due to the nextTick).

Trott commented:
@nodejs/repl This could use some reviews. Any takesr? Even someone blocking it would be more useful than it languishing with no approvals and no requests for changes, IMO.

BridgeAR commented:
Resumed build https://ci.nodejs.org/job/node-test-commit/24000/

BridgeAR commented:
@nodejs/repl @nodejs/tsc PTAL

targos commented:
@BridgeAR can you post an example repl session that shows the impact of this change?

BridgeAR commented:
@targos
> can you post an example repl session that shows the impact of this change?

Before:

```js
process.on('uncaughtException', () => console.log('Uncaught exception'))
throw new Error('foobar')
// Thrown:
// Error: foobar
```

Now:

```js
process.on('uncaughtException', () => console.log('Uncaught exception'))
throw new Error('foobar')
// Uncaught exception
```

CI https://ci.nodejs.org/job/node-test-pull-request/20143/

BridgeAR commented:
Resumed build: https://ci.nodejs.org/job/node-test-commit/25099/

@nodejs/tsc PTAL, this is semver-major and did not get much attention so far.

mcollina commented on doc/api/repl.md:
```diff
@@ -142,13 +142,22 @@ global or scoped variable, the input `fs` will be evaluated on-demand as
-* Uncaught exceptions do not emit the [`'uncaughtException'`][] event.
+* Uncaught exceptions do not emit the [`'uncaughtException'`][] event when
+  creating own `repl` sessions. If such handlers are set in the default `repl`,
+  the error is only passed on to those handlers without further output.
```
I do not understand this. Can you please clarify, maybe with an example?

mcollina commented on lib/repl.js:
```diff
@@ -434,22 +435,31 @@ function REPLServer(prompt,
-    top.displayPrompt();
+    if ((process._forceRepl || isatty(0)) &&
+        process.listenerCount('uncaughtException') !== 0) {
+      process.nextTick(() => {
```
why is this deferred by a nextTick?

BridgeAR replied:
It would otherwise end up in a weird domain hick up: the domain would catch the here emitted exception.

BridgeAR replied:
See also https://github.com/nodejs/node/pull/20803#discussion_r238120518

mhdawson commented:
@BridgeAR can you clarify what question you would like the TSC to answer?

apapirovski commented:
I've removed tsc-review and tsc-agenda labels as this isn't the right use for them — there is no dispute or issue here that can be discussed in the TSC meetings. The best case here is to ping @nodejs/tsc to review the PR.

mhdawson commented:
I think tsc-review is probably applicable as a request for more TSC eyes since it is a SemVer major?

apapirovski commented:
Yeah, that's fair. I'm having a look myself right now.

Fishrock123 commented on doc/api/repl.md:
```diff
@@ -142,16 +142,44 @@ global or scoped variable, the input `fs` will be evaluated on-demand as
 * Trying to use [`process.setUncaughtExceptionCaptureCallback()`][] throws
   an [`ERR_DOMAIN_CANNOT_SET_UNCAUGHT_EXCEPTION_CAPTURE`][] error.
 
+As standalone program:
```
This means `node` without arguments?

BridgeAR replied:
Well, not necessary without arguments. It could include e.g., flags and things like that.

But in the essence, yes (it's the description used in the main repl documentation and I reused it for consistency).

BridgeAR commented:
I just reworked some code here due to @apapirovski request for a test because I realized that adding an `uncaughtException` listener in an `repl` instance which is not run as standalone program is going to silence all main application errors without any chance to get any kind of notification for it.

Therefore I added some more code to fix that behavior as well: it is from now on going to throw an error if the listener is added in such a case. I could theoretically move that part into a separate PR but for me it felt right here.

While writing tests I also noticed that the former way of detecting if the REPL was run as standalone program or not was not sufficient. Currently there's no way to detect if it's the internal repl or not, so I added an internal symbol to detect that as well.

Therefore I'll ask everyone who had a look at this so far to take another look. Thanks.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/20347/

@nodejs/tsc PTAL as this is now also fixing a nasty "bug" that could be triggered when using `uncaughtException` in the repl (silencing the main application errors).

addaleax commented on lib/internal/errors.js:
```diff
@@ -756,6 +756,7 @@ E('ERR_INVALID_CHAR',
     }
     return msg;
   }, TypeError);
+E('ERR_INVALID_COMMAND', '%s', TypeError);
```
So… 
- I feel like this should definitely be a REPL-specific error code if it’s a REPL-specific condition – maybe `ERR_REPL_INVALID_COMMAND`?
- That being said, the condition in which this error is being emitted isn’t even about a command, it’s about a JS statement. How about `ERR_REPL_UNCAUGHT_EXCEPTION_HANDLER`? I don’t think we need to be afraid to get too specific with error codes.

BridgeAR replied:
Hm, I feel most error codes are to specific and my goal here was to explicitly keep it more generic. We could reuse this error code in other places as well (while I do not have a strong opinion about the actual error code and e.g., `ERR_INVALID_STATEMENT` would also be fine for me).

The reason that I prefer less specific error codes is that in most cases users should not have to worry about adding lots of different error codes but to check only for a few which all go down to a similar error condition. That's why I also already thought about adding a kind of grouping property to our errors.

BridgeAR replied:
What I mean is adding something similar like: https://github.com/petkaantonov/core-error-predicates

This module allows user code to catch specific types of errors in a more general fashion instead of checking for each error code specifically. It would e.g., be easier for users to just check for a `FileSystemError` instead of lots of different error codes that relate to file systems when they only want to make sure that a file is written. They might have a fallback mechanism which would just send the file somewhere else or report the error right away.
And even in such cases it would still be possible to check for individual error codes if they occur more frequently.

We also have a couple of errors which could e.g., all be categorized as "user input error". The user would still see what exact error happened but the user would only have to check for a single error type instead of lots of different specific ones.

I hope that makes my intention clearer.

addaleax replied:
Fwiw, there is some existing discussion in https://github.com/nodejs/node/issues/8342 :)

BridgeAR replied:
I switched the error code to `ERR_INVALID_REPL_INPUT`.

addaleax commented on lib/internal/repl.js:
```diff
@@ -61,6 +62,7 @@ function createRepl(env, opts, cb) {
   }
 
   const repl = REPL.start(opts);
+  repl[kIsInternal] = true;
```
I think marking objects as “special” in a certain way is an anti-pattern. Is there a reason not to use a real option?

BridgeAR replied:
The latest version uses the options only and since all functions that use this information are declared inside of the constructor, it's indeed cleaner :+1:.

Trott commented:
Since it's semver-major, should we run CITGM at this time?

addaleax commented on lib/repl.js:
```diff
@@ -104,6 +108,20 @@ for (var n = 0; n < GLOBAL_OBJECT_PROPERTIES.length; n++) {
 
+let tmpListeners = [];
+let checkUncaught = false;
+let currentListeners = [];
```
Why is this global state, and not per REPL instance?

BridgeAR replied:
No REPL instance is allows to add any uncaught exception listeners. If they do, they'll be added to the `currentListeners` array and immediately be removed right after the code has been executed (evaluated). So the single listener is just right and it'll work for `n` REPL intances.

If each REPL instance would track this on their own, we'd add one event listener per instance to track this and that would trigger the event leak warning eventually.

addaleax commented:
> Since it's semver-major, should we run CITGM at this time?

I don’t think we have modules in CITGM that test the default REPL (aka Node’s “interactive” mode) in any way, so I’m not sure it’s worth it?

BridgeAR commented on doc/api/errors.md:
```diff
@@ -1207,6 +1207,11 @@ A callback function was required but was not been provided to a Node.js API.
+<a id="ERR_INVALID_COMMAND"></a>
+### ERR_INVALID_COMMAND
+
+The command (likely a REPL command) can not be used in the specific context.
```
I am open to suggestions and I'll try to think about a better wording as well. Right now I can't think about much but I agree that it should be improved.

BridgeAR replied:
I rewrote the description.

BridgeAR commented:
I updated the PR and incorporated the feedback. I also added a `known issue` test that highlights one of the problems this PR partially fixes (using the `uncaughtException` listener from an REPL instance silences all errors in the application that started that instance).

PTAL

richardlau commented on test/known_issues/known_issues.status:
```diff
@@ -5,6 +5,7 @@ prefix known_issues
 # sample-test                        : SKIP
 
 [true] # This section applies to all platforms
+test-repl-uncaught-exception-error: SKIP
```
Why is this skipped everywhere? If it passes then it shouldn't be in known issues.

BridgeAR replied:
Please check the comments inside of the test (https://github.com/nodejs/node/pull/20803/files#diff-4f5313e8a4583fdc4637c1a56958e45fR7).

Thrown errors are suppressed even though they should not be. I could not think of a way to write the test that it would fail and still show the mentioned issue.

richardlau replied:
But marking the test skipped everywhere means that this is dead code that isn't executed.

BridgeAR replied:
I know... :/ do you have a suggestion what to do instead? I would really like to have a known issue test for this. IMO this is a pretty nasty error.

BridgeAR replied:
I could move the test to `parallel`. It would be an `inverse` known issue... It would fail in case we fix the issue.

richardlau replied:
> 
> 
> I know... :/ do you have a suggestion what to do instead? I would really like to have a known issue test for this. IMO this is a pretty nasty error.

cc @nodejs/testing

BridgeAR replied:
I fixed it by using a child process. That way I verify if the process kept on running after the exception or not.

It will now always fail since it keeps on running.

BridgeAR commented on test/parallel/test-repl-pretty-custom-stack.js:
```diff
@@ -40,8 +39,6 @@ process.on('uncaughtException', (e) => {
   throw e;
 });
 
-process.on('exit', () => (Error.prepareStackTrace = origPrepareStackTrace));
```
It is somewhat unrelated. In a former PR this test was impacted as well and I noticed that this is actually not necessary, so I just went ahead and removed it.

BridgeAR commented on doc/api/repl.md:
```diff
@@ -138,16 +138,47 @@ global or scoped variable, the input `fs` will be evaluated on-demand as
-* Uncaught exceptions do not emit the [`'uncaughtException'`][] event.
+* Uncaught exceptions only emit the [`'uncaughtException'`][] event if the
+  `repl` is used as standalone program. If the `repl` is included anywhere in
+  another application, adding this event synchronous will throw an
```
This is actually outdated in the other PR. In this version I was not able to actually handle listeners that were added asynchronously. E.g., `setTimeout(() => process.on("uncaughtException", () => {}))` was not detected here. This is fixed in the latest implementation.

BridgeAR commented on doc/api/errors.md:
```diff
@@ -1308,8 +1308,14 @@ An invalid `options.protocol` was passed.
+<a id="ERR_INVALID_REPL_INPUT"></a>
+### ERR_INVALID_REPL_INPUT
+
+The input can or may not be used in the [`REPL`][]. All prohibited inputs are
```
I thought I'll keep it broad so that it could be used for different REPL input. So far it will only be used for uncaught exception listeners, so only the `may` part applies. I am happy to change it to only `may` or what ever you have as alternative.
